### ISSUE TITLE
Sorting based on nested object key is not working

### ISSUE THREAD

**Issue by sajidfrommerqata** [mui-org/material-ui-x#1363]
In the below code, inside renderCell you can  see I'm getting branch_location from business_branch object and business_branch object is inside the order object. 

**Issue:** The issue is when I do so, The column sort feature stops working

```
{
    field: "branch_location",
    headerName: "Vendor",
    width: 170,
    type: "string",
    renderCell: (params) => params.row.order.business_branch.branch_location,
  },
```

Expected Result: Sorting should work even If I read values from nested object.

**Comment by rexwreyes** [comment]
I am having what i think is the same problem....

`{ field: 'subcategory', headerName: 'Subcategory', flex: 1, valueFormatter: ({ value }) => value.name, type: 'string' }`

with JSON that looks like...
`{   name: "Wine (11oz)"
    shortname: "Wine"
    subcategory: { name: "Wine Glassware" }
}`

Other columns sort and filter fine.  With this column it sorts to something obviously wrong and filtering shows no content

**Comment by oliviertassinari** [comment]
@rexwreyes do you have a reproduction?

**Comment by rexwreyes** [comment]
Cant really share links, i'll try to get a CodeSandbox built

**Comment by rexwreyes** [comment]
Well i can't seem to import DataGrid to the template you provided above

`DependencyNotFoundError
Could not find dependency: '@material-ui/data-grid' relative to '/src/Demo.tsx'`

**Comment by rexwreyes** [comment]
@oliviertassinari https://codesandbox.io/s/no-sort-krk0x

**Comment by oliviertassinari** [comment]
@rexwreyes Interesting, it's an issue in the domain of the API/documentation/missing warning.

It seems that we have an opportunity to add a new warning when the configuration of the column definitions is wrong.

**Comment by dtassone** [comment]
@rexwreyes yes you need to implement a `sortComparator` to let the grid knows how to compare the data if you use renderCell. Otherwise you can use `valueGetter` to get the data, then the built-in string comparator should work.

**Comment by oliviertassinari** [comment]
I'm reopening so we can keep exploring the path that would yield to the introduction of a warning. Future complaints from the developers will help us determine its priority.

AFAIK, sorting and filtering fails because the component tries to apply a string column type methods to an object input, which can be easily detected.

**Comment by rexwreyes** [comment]
For anyone else who had this problem.  the fix i found was changing this...`{ field: 'subcategory', headerName: 'Subcategory', flex: 1, valueFormatter: ({ value }) => value.name, type: 'string' }` to this....`{ field: 'subcategory', headerName: 'Subcategory', flex: 1, valueGetter: (params) => {return params.row.subcategory.name}}`

**Comment by sajidfrommerqata** [comment]
> For anyone else who had this problem. the fix i found was changing this...`
> 
> ```diff
> {
>   field: 'subcategory',
>   headerName: 'Subcategory',
>   flex: 1,
> - valueFormatter: ({ value }) => value.name,
> + valueGetter: (params) => params.row.subcategory.name,
>   type: 'string'
> }
> ```

But what if the value is wrapped between HTML tags
{
  field: 'subcategory',
  headerName: 'Subcategory',
  flex: 1,
- valueFormatter: ({ value }) => value.name,
+ valueGetter: (params) => <Chip size="small" className="Mu-status mu-red" _**label={params.row.subcategory.name}**_ />,
  type: 'string'
}

**Comment by oliviertassinari** [comment]
@sajidfrommerqata This looks like another opportunity to throw a warning. valueGetter can't return a React element.

**Comment by sajidfrommerqata** [comment]
@oliviertassinari the serious issue which I'm facing in my project is Sorting. I'm unable to sort the column if the value is wrapped in some HTML tags or React Elements

**Comment by dtassone** [comment]
you should use a render cell to render react elements and valueGetter to get the correct value.

**Comment by mikedidomizio** [comment]
> But what if the value is wrapped between HTML tags. You can see below code that the value is assigned to the label attribute of the chip
> 
> ```diff
> {
>   field: 'subcategory',
>   headerName: 'Subcategory',
>   flex: 1,
> - valueFormatter: ({ value }) => value.name,
> + valueGetter: (params) => <Chip size="small" className="Mu-status mu-red" label={params.row.subcategory.name} />,
>   type: 'string'
> }
> ```

You can do it this way (as mentioned by @dtassone )

```
     valueGetter: (params: GridValueGetterParams) => params.row.object.value, // this is the filtering value
     renderCell: (params: GridValueGetterParams) => {  // inside is what's rendered in the cell
```

Tested with

```
    "@material-ui/core": "^4.9.12",
    "@material-ui/data-grid": "^4.0.0-alpha.31",
    "@material-ui/icons": "^4.11.2",
```

**Comment by m4theushw** [comment]
To give an update on the status of https://github.com/mui-org/material-ui-x/issues/1363#issuecomment-813985262. 

- The [filtering page](https://mui.com/components/data-grid/filtering/#predefined-filters) already mentions which value is used for filtering
- The description of [`valueFormatter`](https://mui.com/components/data-grid/columns/#value-formatter) tells it's only used during rendering, but the message is kinda hidden. It could be formatted as a warning
- The [sorting page](https://mui.com/components/data-grid/sorting/#basic-sorting) doesn't mention which value is used for. It could have a similar description to the one in the filtering page.
- We still miss the runtime warning when `valueGetter` returns an element.
